demos/link-prediction/random-walks/ctdne-link-prediction.ipynb

```diff
@@ -0,0 +1,680 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def create_link_examples(g, edges):\n",
+    "    pos = [(src, tgt) for (src, tgt) in edges[[\"source\", \"target\"]].values]\n",
+    "    neg = sample_negative_examples(g, pos)\n",
+    "    return pos, neg\n",
+    "\n",
+    "\n",
+    "def sample_negative_examples(g, positive_examples):\n",
+    "    possible_edges_all = {\n",
+    "        (src, tgt) for src in g.nodes() for tgt in g.nodes() if src != tgt\n",
+    "    }\n",
+    "    possible_edges_neg = list(\n",
+    "        possible_edges_all.difference(\n",
+    "            {e for (src, tgt) in positive_examples for e in [(src, tgt), (tgt, src)]}\n",
+    "        )\n",
+    "    )\n",
+    "    return random.choices(possible_edges_neg, k=len(positive_examples))\n",
+    "\n",
+    "\n",
+    "pos, neg = create_link_examples(graph, edges_train)\n",
+    "pos_test, neg_test = create_link_examples(graph, edges_test)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "It's good to verify that the data structures we've created so far from the raw data look reasonable.\n",
+    "\n",
+    "`StellarGraph.info` is a useful method for inspecting the graph we've created, and we can also check that the number of link examples correspond to the train/test subset values we defined earlier."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "StellarGraph: Undirected multigraph\n",
+      " Nodes: 151, Edges: 37929\n",
+      "\n",
+      " Node types:\n",
+      "  default: [151]\n",
+      "    Features: none\n",
+      "    Edge types: default-default->default\n",
+      "\n",
+      " Edge types:\n",
+      "    default-default->default: [37929]\n",
+      "\n",
+      "Training examples: 9524 positive links, 9524 negative links\n",
+      "Validation examples: 3119 positive links, 3119 negative links\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(graph.info())\n",
+    "print()\n",
+    "print(f\"Training examples: {len(pos)} positive links, {len(neg)} negative links\")\n",
+    "print(\n",
+    "    f\"Validation examples: {len(pos_test)} positive links, {len(neg_test)} negative links\"\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Running random walks\n",
+    "\n",
+    "We try to keep the setup comparable between the use of temporal and biased (static) random walks. For temporal walks, the input parameter is defined in terms of the total number of context windows you are interested in obtaining, which differs from the traditional approach of specifying the number of walks to run per node in the graph. We calculate the number of context windows we need in terms of the traditional parameters as:\n",
+    "```\n",
+    "num_cw = num_nodes * num_walks_per_node * (walk_length - cw_size)\n",
```

Comment:
What's the motivation for repeating the code here? If you think this is clearer, maybe the code could be tweaked to allow writing this directly? E.g.

> Define the parameters we use ...
> ```python
> num_walks_per_node = 10
> walk_length = 80
> cw_size = 10
> ```
> We try to keep the setup comparable [...] We calculate the number of context windows we need in terms of the traditional parameters as:
>
> ```python
> num_cw = graph.number_of_nodes() * num_walks_per_node * (walk_length - cw_size)
> ```
>
> We're now ready to do the walks:
>
> ```python
> temporal_rw = TemporalRandomWalk(graph)
> ...
> ```

The discussion of what was recommend in the reference paper could then move into the discussion of the `num_cw` cell, too.

Reply:
What I initially wanted to do was to show the equation from the paper in the markdown, and then write a comment on that line of the code to indicate that that expression matches the equation, but yeah it looks a bit odd that it's just repeating the same code.

I'll split into multiple cells like you've suggested